### ISSUE TITLE
exiv2 0.27.3: output truncation leads to misleading data

### ISSUE THREAD

**Issue by LebedevRI** [Exiv2/exiv2#1380]
Example:
[bug.zip](https://github.com/Exiv2/exiv2/files/5447303/bug.zip)
```
$ exiv2 -u -PEIXxkycsvt bug.dng | grep -i Exif.Image.ColorMatrix2
Error: Directory Nikon1 with 25665 entries considered invalid; not read.
0xc622 Exif.Image.ColorMatrix2                      SRational   9  72  8695/10000 -2558/10000 -648/10000 -5015/10000 12711/10000 2575/10000 -1279/10000 2215/10000 7514/100 ...  8695/10000 -2558/10000 -648/10000 -5015/10000 12711/10000 2575/10000 -1279/10000 2215/10000 7514/100 ...
```
The last element in color matrix appears to be `7514/100`, but it's actually `7514/10000`, and the output is somehow truncated.

This e.g. breaks darktable's https://github.com/darktable-org/darktable/blob/master/tools/dngmeta.sh script.

**Comment by clanmills** [comment]
Good to hear from you.  Long time no speak.

This due to truncation of the ascii string reporting the data.  The program exiv2 is truncates your data, however exifprint does not.  So the API is presented with the correct data and it's been "chopped" by the exiv2 command-line program.

```bash
889 rmills@rmillsmbp:~/gnu/exiv2/team/book/build $ exiv2 --grep ColorMatrix2/i ~/Downloads/bug.dng 
Exif.Image.ColorMatrix2                      SRational   9  8695/10000 -2558/10000 -648/10000 -5015/10000 12711/10000 2575/10000 -1279/10000 2215/10000 7514/100 ...  <--- note the ...
890 rmills@rmillsmbp:~/gnu/exiv2/team/book/build $ exifprint ~/Downloads/bug.dng | grep ColorMatrix2 
Exif.Image.ColorMatrix2                      0xc622 SRational   9  8695/10000 -2558/10000 -648/10000 -5015/10000 12711/10000 2575/10000 -1279/10000 2215/10000 7514/10000
892 rmills@rmillsmbp:~/gnu/exiv2/team/book/build $ 
```

The exiv2 command option `--binary` disables truncation of long strings. 

```bash
902 rmills@rmillsmbp:~/gnu/exiv2/team/book/build $ exiv2 -u --binary -pv -g ColorMatrix2 ~/Downloads/bug.dng 
0xc622 Image        ColorMatrix2                SRational   9  8695/10000 -2558/10000 -648/10000 -5015/10000 12711/10000 2575/10000 -1279/10000 2215/10000 7514/10000
903 rmills@rmillsmbp:~/gnu/exiv2/team/book/build $ 
```

If you're happy with this explanation, I invite you to close the issue.  If you want to discuss this further, I will be happy to hear your comments.

-------------


Incidentally (this is irrelevant, however interesting to me) the code in my book also finds and reports 9 SRATIONALs.  The code in my book is an "all new" metadata read engine and not based on Exiv2.

```bash
892 rmills@rmillsmbp:~/gnu/exiv2/team/book/build $ ./tvisitor -pRU ~/Downloads/bug.dng | grep c622 
     358 | 0xc622 Exif.Image.0xc622            | SRATIONAL |        9 |      9546 | 8695/10000 4294964738/10000 4294966648/10000 4294962281/10000 12711/10000 2575/10000 4294966017/10000 2215/10000 7514/10000
893 rmills@rmillsmbp:~/gnu/exiv2/team/book/build $ 
```

**Comment by clanmills** [comment]
I'm going to close this.  If you wish to further discuss more, I'll be happy to reopen this.

**Comment by LebedevRI** [comment]
> The exiv2 command option `--binary` disables truncation of long output strings.

It is good to know that there is a workaround for this bug.
I'm guessing you won't be open to fixing it?
Can you at least reference the change that introduced it, for completeness?

**Comment by clanmills** [comment]
I don't believe it is a bug.   I believe you have to add `--binary` option to the scripts when you call exiv2.

I don't recall the history of the `--binary` option.  Perhaps exiv2 used to default to that and the script was good.  However, I suspect that it's always been this way and nobody noticed that the script wasn't working correctly when the output was long.

**Comment by clanmills** [comment]
You're right.  I did it.  Guilty!  I can't remember what `--binary` did before I made that change.

Can you try adding `--binary` to your scripts.  I'll reopen this and investigate the "old" behaviour.

```
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  609)     static void binaryOutput(bool suppressLong,const std::ostringstream& os)
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  610)     {
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  611)         const int dots = 100;
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  612)         if ( suppressLong && os.str().length() > dots ) {
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  613)            std::cout << os.str().substr(0,dots) << " ..." ;
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  614)         } else {
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  615)             std::cout << os.str();
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  616)         }
7515d85a0 (clanmills      2020-05-12 10:58:29 +0100  617)     }
```

**Comment by clanmills** [comment]
I believe I made this change because "pure binary" output was causing chaos in the test suite.  It's a bad idea to write binary bytes to std::cout.  The API presents binary data correctly.  The `--binary` option is only used in a couple of locations to output `(Binary output suppressed)` when there are more than 128 bytes.

```cpp
            if (Params::instance().binary_ &&
                (md.typeId() == Exiv2::undefined || md.typeId() == Exiv2::unsignedByte ||
                 md.typeId() == Exiv2::signedByte) &&
                md.size() > 128) {
                std::cout << _("(Binary value suppressed)") << std::endl;
                return true;
            }
```

So, I hijacked/refocussed the `--binary` option to truncate the output of long lines with the ascii ellipsis `...` (3 ascii periods).

I don't want to restore the "old" behaviour because that will restore the chaos.  I think the best fix is to output long lines.  The option `--binary` will be totally ignored and the function binaryOutput() becomes:

```cpp
static void binaryOutput(bool /*suppressLong*/,const std::ostringstream& os)
{
    std::cout << os.str();
}
```

What do you think?

**Comment by LebedevRI** [comment]
Thank you for double-checking!
I don't have an opinion on printing truly binary data - `(Binary value suppressed)` is probably always fine there.
So as long as that restores/fixes/makes `Exif.Image.ColorMatrix2` print "correctly" again, i'll be happy.

**Comment by clanmills** [comment]
I'll do something next week.  I'll leave this for a couple of days to see if I have an inspired idea.

Are you willing to add `--binary` to the exiv2 commands in your script?

**Comment by clanmills** [comment]
I'm submitted a PR to fix this:  #1384.

Effectively, the option `--binary` does nothing!  You can add option --binary to your scripts and DT will fixed without waiting for a new release of Exiv2.  Please let me know if you decide to do that. 

I haven't updated the man page exiv2.1 yet.
